Add ppc64le and bump Python versions in Travis CI

peternowee opened the pull request:
<details><summary>Outdated description of outdated commit 2d60c6ed. Updated description in <a href="https://github.com/pydot/pydot/pull/245#issuecomment-727197060">my comment of 2020-11-14 below</a>.</summary>
Considering that:

- IBM, through its supplier Datamatics, requested many Python projects to [add the `ppc64le` (Power) architecture][1] to their Travis CI test matrices. This request was also made to pydot in PR [pydot/pydot#243][2] by GitHub user @asellappen.
- Some related links:
  - ibm.com: [Build your open source projects on IBM Power Systems with Travis CI][3].
  - youtube.com: [OpenPOWER Summit NA 2019: Developing Applications using Acceleration Hardware on Power][4].
- Gerrit Huizenga of IBM (@gerrith3 on GitHub) said he thinks this will not lead to additional billing from Travis CI to the OSS community, because [he provides Travis CI with the Power hardware][5].
- The [Travis CI documentation on CPU architectures][6] currently says IBM architecture support is in beta stage and only available for Open Source repositories.
- There are many abstraction layers between pydot and the hardware architecture, including other libraries, Python and the operating system. Pydot itself does not need to be compiled for a specific architecture. In a perfect world, pydot is architecture-independent and testing pydot on a single architecture should be sufficient.
- However, the world is not perfect and [abstraction layers can be leaky][7]. According to Gerrit Huizenga of IBM, [corner cases do come up][8], which is why they do so much testing up front and provide free access to testing.
- Still, even if an automated testing service is provided for free, it has an [environmental footprint][9], so there are limits to how much we should make use of that as well.
- Travis CI also requests open-source users to remember that Travis CI provides its service free of charge to the community and only specify the matrix we [actually need][10].

After weighing all of the above, I prefer not to extend the number of builds by [matrix expansion][11] as proposed by IBM, because that would multiply the number of builds by a factor of 2: 5 Python versions * 2 architectures = 10 builds of 2 to 3 minutes each.

Instead, I propose we list individual build jobs based on the following two ideas:

- We only build for all architectures for the lowest and the highest Python versions.
- We only test the Python versions in-between on one architecture each, though that does not necessarily have to be amd64. Preferably, we just step diagonally between the different architectures and the in-between Python versions.

This way, we only increase the number of automated Travis CI builds from 5 to 7, so by 40% instead of the 100% increase of the original request.

We can always reconsider this policy if some real examples come up of pydot bugs specific to a combination of an in-between Python version and a specific architecture.

[1]: https://github.com/pulls?q=author%3Aasellappen+ppc64le+travis
[2]: https://github.com/pydot/pydot/pull/243
[3]: https://developer.ibm.com/tutorials/travis-ci-on-power/
[4]: https://www.youtube.com/watch?v=qPZtf3hoMSs
[5]: https://github.com/nrdcg/goinwx/pull/5#issuecomment-721422838
[6]: https://docs.travis-ci.com/user/multi-cpu-architectures/
[7]: https://www.joelonsoftware.com/2002/11/11/the-law-of-leaky-abstractions/
[8]: https://github.com/nrdcg/goinwx/pull/5#issuecomment-719084047
[9]: https://en.wikipedia.org/wiki/Data_center#Energy_use
[10]: https://docs.travis-ci.com/user/build-matrix/#listing-individual-jobs
[11]: https://docs.travis-ci.com/user/build-matrix/#matrix-expansion
</details>

@asellappen and @gerrith3: Let me know what you think of this instead of pydot/pydot#243.
@ldez: Thanks for asking in https://github.com/nrdcg/goinwx/pull/5 many of the questions I also had.

The other commit in this PR removes the workaround for the Python 3.7 test environment added by commit eb7aef13. It is no longer needed by travis-ci.com.

prmtl commented:
I disagree with the "diagonal approach" to testing. We should keep amd64 as a baseline and not skip any currently supported Python version (3.6-3.9 right now + 2.7 until pydot drop it, so 10 jobs). You will lower carbon/env. footprint on CI but increase it in anyone that will have issues on Python 3.X + amd64 that will try to debug it (yourself included) in case of problems. For ppc64le we also should do it like that or just ask @asellappen and @gerrith3 which versions they actually need - otherwise what is the point?

Alternatively we can do 2.7  + lowest (3.6) + highest (3.9) on both architectures (so 6 jobs)

peternowee commented:
I just do not really believe we will soon find a pydot bug that is specific to a combination of a specific architecture *and* a specific Python version. If we find some problem on ppc64le, it is highly likely it will come out on several Python versions. If there was a specific ppc64le bug in a specific Python version, it would also probably already come out during testing of Python. So, I don't think the energy saving in all those many CI runs will be offset by the few times someone might actually have something to debug his way to such a specific bug. I would really like to see those cases first. Otherwise we could just as well keep multiplying in all dimensions for all the possible combinations of circumstances, it can be endless.

I don't mind dropping the diagonal testing, just moving all the in-between Python versions back to amd64. So amd64 for all, and ppc64le only for lowest Python (currently 2.7) and the highest (3.7 or 3.9, but that also depends on what the Travis environment has installed, see https://github.com/mcmtroffaes/pybtex-docutils/commit/a521cba9694e95210c4f0c6f4848b70c1fb1d742 for example). I am also ok with letting them chose 2 Python versions. Just but not all of them.

Again, we are several abstraction layers away from the architecture and I am really bothered by the environmental impact of all these automated CIs that get flipped on so easily and then never turned off again. They may be "free" to us money-wise, but we all read about the environmental impact in the newspapers every day. Here is our chance to do a little thing about it ourselves.

peternowee commented:
@gerrith3 Thanks for your understanding.

@prmtl @gerrith3 How about 2.7+3.6+3.9 for amd64 and 2.7+3.9 for ppc64le. Again, I really do not want to just double.

prmtl commented:
> @prmtl @gerrith3 How about 2.7+3.6+3.9 for amd64 and 2.7+3.9 for ppc64le. Again, I really do not want to just double.

Cool.

peternowee commented:
> > @prmtl @gerrith3 How about 2.7+3.6+3.9 for amd64 and 2.7+3.9 for ppc64le. Again, I really do not want to just double.
> 
> Cool.

@gerrith3 Also ok for you?

gerrith3 commented:
@peternowee Yep - I think that is a nice compromise and should catch any corner cases.  Thanks again for the extended thought on this one.  We are trying to do minimal changes in general but really appreciate the additional thought to validate the architecture independence in a resource conserving way.

peternowee commented:
@prmtl @gerrith3 Ok, thank you both as well. I will check what is available on Travis exactly and then replace the commit here, probably tomorrow.

gerrith3 commented:
I don't think this is a problem here, but remember if you need more travis credits under the new policy you can either ask via support@travis-ci.com or via this support ticket link: https://support.travis-ci.com/.  They posted a broad blog about the changes but a lot of people are thinking it changes open source support but they are cracking down on bitcoin-miners - as long as you have an authenticated ID and open source repositories you qualify for as many credits as you need.

peternowee commented:
@prmtl @gerrith3 @asellappen Ok, I updated the main commit. I rewrote parts of the commit message and I also added some comments to the `.travis.yml` file to explain the new structure. Let me know if there are any further remarks, otherwise I will merge it in a few days.

#### Add ppc64le and bump Python versions in Travis CI

Considering that:

- IBM, through its supplier Datamatics, requested many Python projects to [add the `ppc64le` (Power) architecture][1] to their Travis CI test matrices. Related links:
  - ibm.com: [Build your open source projects on IBM Power Systems with Travis CI][2].
  - youtube.com: [OpenPOWER Summit NA 2019: Developing Applications using Acceleration Hardware on Power][3].
  - Gerrit Huizenga of IBM (@gerrith3 on GitHub) provided more details on the [motives and open source contributions of IBM][4] in answer to questions from Ludovic Fernandez (@ldez) of nrdcg/goinwx.
- The request to add `ppc64` testing was also made to pydot in PR [pydot/pydot#243][5] by @asellappen of Datamatics. The proposed implementation extended the test job configuration by [matrix expansion][6].
- [Matrix expansion][7] means adding a new dimension to the build configuration (e.g. `arch`, `os`, `env`), after which the number of test jobs is automatically multiplied by the number of choices given in the newly added dimension. In this case, it would add a dimension of two architectures, expanding the matrix of test jobs by a factor of 2: 5 Python versions (at that time) * 2 architectures = 10 test runs of 2 to 3 minutes each.
- Gerrit Huizenga of IBM said he thinks the adding of `ppc64le` test jobs to Travis will not lead to additional billing from Travis CI to the OSS community, because [he provides Travis CI with the Power hardware][8].
- The [Travis CI documentation on CPU architectures][9] currently says IBM architecture support is in beta stage and only available for Open Source repositories.
- There are many abstraction layers between pydot and the hardware architecture, including other libraries, Python and the operating system. Pydot itself does not need to be compiled for a specific architecture. In a perfect world, pydot is architecture-independent and testing pydot on a single architecture should be sufficient.
- However, the world is not perfect and [abstraction layers can be leaky][10]. According to Gerrit Huizenga of IBM, [corner cases do come up][11], which is why they do so much testing up front and provide free access to testing.
- Still, even if an automated testing service is provided for free, it has an [environmental footprint][12], so there are limits to how much we should make use of that as well.
- Travis CI also requests open-source users to remember that Travis CI provides its service free of charge to the community and only specify the matrix we [actually need][13].
- Even if we ever find an architecture-specific bug in pydot, it seems highly likely that it will come out on several Python versions for that architecture.
- All of this was further discussed between pydot maintainers Sebastian Kalinowski (@prmtl) and Peter Nowee (@peternowee) and Gerrit Huizenga of IBM in PR [pydot/pydot#245][14].

The following was decided:

- Baseline testing is kept at `amd64`, with testing of the lowest and the highest Python versions supported by the Python community, as well as the lowest end-of-life (EOL) Python version that we still choose to support pydot on. Currently, that means Python 3.6, 3.9 and 2.7 respectively.
- For other architectures, such as `ppc64le`, we do limited testing to catch reasonably expectable corner cases and validate architecture independence in a resource conserving way. For now, we start with Python 2.7 and 3.9.

We can always reconsider this policy depending on actual findings.

[1]: https://github.com/pulls?q=author%3Aasellappen+ppc64le+travis
[2]: https://developer.ibm.com/tutorials/travis-ci-on-power/
[3]: https://www.youtube.com/watch?v=qPZtf3hoMSs
[4]: https://github.com/nrdcg/goinwx/pull/5
[5]: https://github.com/pydot/pydot/pull/243
[6]: https://docs.travis-ci.com/user/build-matrix/#matrix-expansion
[7]: https://config.travis-ci.com/matrix_expansion
[8]: https://github.com/nrdcg/goinwx/pull/5#issuecomment-721422838
[9]: https://docs.travis-ci.com/user/multi-cpu-architectures/
[10]: https://www.joelonsoftware.com/2002/11/11/the-law-of-leaky-abstractions/
[11]: https://github.com/nrdcg/goinwx/pull/5#issuecomment-719084047
[12]: https://en.wikipedia.org/wiki/Data_center#Energy_use
[13]: https://docs.travis-ci.com/user/build-matrix/#listing-individual-jobs
[14]: https://github.com/pydot/pydot/pull/245

asellappen commented:
I am fine for this change

peternowee commented:
@asellappen Ok, thank you and good luck with your project.

asellappen commented:
> @asellappen Ok, thank you and good luck with your project.

Thank you so much for taking care of this and wishes....

gerrith3 commented:
Excellent write up (again!) and I'm good with the well structured thinking and recording of the path to get here. Thanks, @peternowee !!
